### PR TITLE
Refactor configuration handling

### PR DESCRIPTION
**Pull request by Darth-Mera**
This is a work in progress on refactoring the handling of configuration. 

I have decided to scratch my first solution since it grew unnecessarily bloated and come up with something hopefully a bit more elegant. 

Here I am: 
-using pflag package to have POSIX/GNU-style flags, both long and short versions
-using Viper to introduce support for more config formats (I've played around with JSON and YAML, but it should also support TOML and some others) . Have to recognize that Viper is a big package, so if this is not desired, it should be quick and easy to revert back to the custom JSON unmarshalling, but my personal opinion is that having support of more formats is great.
-proposing a change to the config structure so it's a bit more human-eye friendly 

I have moved all the config handling logic to its separate package, exporting just a New() function that is called from main() to create a new config instance. 

My plan is to also move DB connection logic there, and just pass pointers around the program. 

I've also moved the global constants to its own package to avoid circular dependencies. 

As it is WIP, there's currently only little validation or tests, I'm working on adding them. 

PS: I realise that the DB logic is subject to change as Lev is implementing the multiple DB support - it will be changed accordingly later.

### CODE REVIEW

**Comment by lveyde** [comment]
Haven't reviewed it completely, since it's marked WIP, but I think that it's better to use flag package instead, as not to change the command line interface.

**Comment by alessandrocarminati** [comment]
My gut feelings are that having all these dependencies added for having the configuration in many formats is a bit more overkill.
I don't think this tool will ever be certified for FUSA. It is helping FUSA experts for their analysis but since it is written in GOLANG, which comes with a non neglectable runtime, its certification is difficult.
That been said, adding a fat dependency as viper is just for having the configuration possible with different formats does not add anything to its functionality and complicates its code base.
For the same reason, I see adding dependencies for the flags parsing with caution, but I do see benefits in having more articulated parsing with respect to the one I made from scratches.

**Comment by alessandrocarminati** [comment]
In addition to this, obviously, but it is better to state it clearly, I found it very useful to move the configuration logic in a package so that it can be reused in the other tool. Finally, the tests are more than welcome.
This topic was already discussed on other channels, but for transparency sake, it is better to write this here.

File: config/config.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+	if err := v.ReadInConfig(); err != nil {
+		return fmt.Errorf("problem while reading config file")
+	}
+	if err := v.Unmarshal(&cfg); err != nil {
+		return fmt.Errorf("problem while parsing config file: %w", err)
+	}
+	return nil
+}
+
+func (cfg *ConfValues) validate() error {
```
Comment (rokkbert) [inline config/config.go:100]: When we support multiple databases with different options the validation of them will be more complicated. Or we just don't validate those options. If the DB-connection is done early enough there's not so much value in validating the config. A good error message in case of connection failures is sufficient.

File: config/flags.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	fs.StringSliceP("excluded-before", "b", nil, "list of `symbols` to exclude before the target symbol")
+	fs.StringSliceP("excluded-after", "a", nil, "list of `symbols` to exclude after the target symbol")
+	fs.StringSliceP("target-subsys", "t", nil, "list of `subsystems` to include in the output")
+
+	fs.StringP("db-name", "n", "", "target database `name`")
+	fs.StringP("db-url", "d", "", "database `URL`")
+	fs.StringP("db-user", "u", "", "database `user`")
+	fs.StringP("db-password", "p", "", "database `password`")
+	fs.IntP("db-port", "P", 0, "database `port`")
+	fs.IntP("db-instance", "i", 0, "database `instance`")
```
Comment (rokkbert) [inline config/flags.go]: How is this related to the DB? Isn't it "The interesting symbols instance identifier", like the original README.md says?

File: config/flags.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+func parseCommandLine(fs *pflag.FlagSet, configPath *string, args []string) error {
+	fs.SortFlags = false // disable sorting so flags get printed in the order they are defined
+	initFlagSet(fs, configPath)
+
+	if err := fs.Parse(args); err != nil {
+		return fmt.Errorf("unable to parse args: %w", err)
+	}
+
+	help, err := fs.GetBool("help")
+	if err != nil {
```
Comment (rokkbert) [inline config/flags.go:42]: Since this is a BoolP I'm not sure what error could possibly occur. The user doesn't say "-h true|false" here, after all. Is this intended to catch bugs, when we accidentally remove the flag form the FlagSet or something like that or just based on the general principle of not ignoring error values?

File: config.go [hunk 4]
```diff
@@ -5,11 +5,13 @@
 
 package main
 
+/*
```
Comment (rokkbert) [inline config.go]: Is this file kept for reference until the PR is finished?

File: config/config.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
+package config
+
+import (
+	"fmt"
+
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
```
Comment (rokkbert) [inline config/config.go]: I don't think that more config file formats are a good thing. I believe it's the opposite, because it just brings in more code and confusion, when everybody uses different files, which are harder to compare. I don't see the need for that. All the formats are, for our purposes here, simple enough so that everybody can learn to use them and personal tastes shouldn't matter. I didn't even know, for example, something like TOML exists and I have no idea why somebody thought the world needs another one of those

File: config/config.go [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+	confValues, err := initConfig()
+	config.ConfValues = confValues
+	if err != nil {
+		return nil, err
+	}
+	return &config, nil
+}
+
+// initConfig gathers the configuration values from command line and config file (if provided) and returns
+// them in a ConfValues struct.
```
Comment (rokkbert) [inline config/config.go]: Do command line arguments override those from the config file?

**Comment by rokkbert** [comment]
> Haven't reviewed it completely, since it's marked WIP, but I think that it's better to use flag package instead, as not to change the command line interface.

But was it changed? Weren't there just things added for which no flags existed yet?

**Comment by rokkbert** [comment]
> In addition to this, obviously, but it is better to state it clearly, I found it very useful to move the configuration logic in a package so that it can be reused in the other tool. Finally, the tests are more than welcome. This topic was already discussed on other channels, but for transparency sake, it is better to write this here.

After looking at the config here and in kern_bin_db I'm not sure how much reuse is possible, but separate package is cleaner in any case.

**Comment by rokkbert** [comment]
An important help/validation IMHO would be to tell the user exactly which mandatory arguments he's missing, instead of just "missing needed arg". I think that would fit well in here.

File: config.go (hunk 4 above)
Comment (Darth-Mera) [reply config.go]: Yes

File: config/config.go (hunk 6 above)
Comment (Darth-Mera) [reply config/config.go]: Yes, that's the desired behavior AFAIK

File: config/config.go (hunk 5 above)
Comment (Darth-Mera) [reply config/config.go]: I will roll this back and remove the viper dep

File: config/flags.go (hunk 3 above)
Comment (Darth-Mera) [reply config/flags.go:42]: My thinking was more around the general "do not ignore errors" principle :)

File: config/flags.go (hunk 2 above)
Comment (Darth-Mera) [reply config/flags.go]: This was cleared on Slack, so I'm marking the conversation as resolved.

**Comment by Darth-Mera** [comment]
I have removed the Viper dep and rolled back to just supporting JSON config. 
However I would argue for keeping the pflag implementation, as it IMHO allows to implement what we want with less code and it's more readable. 

As of now, this implementation works, but I will keep the PR as a WIP until the changes for multiple DB support are approved and I can implement them here as well.

**Comment by alessandrocarminati** [comment]
Sorry for the delay in reviewing your contribution, we had a few other things that needed our attention first.

Just wanted to give you a heads-up that I merged odra's contribution this morning, which conflicts with yours. Before I dive into the review, could you please rebase your work to the latest commit? Also, for future, it would be awesome if you could provide a squashed commit.

Thanks a lot for your help.

**Comment by Darth-Mera** [comment]
@alessandrocarminati  

I can squash it to one commit, no problem, I have an experience that for reviewing it's better to see the chain of changes/commits, so I left the history, but I can clean it up. 

Re rebasing: I've written my tests with testify, but the changes you've merged use ginkgo framework. So it seems there's some decision needed on the way we're supposed to test this tool.

File: README.md [hunk 7]
```diff
@@ -42,46 +42,53 @@ $ ./nav -f conf.json -s kernel_init
...
+  -f, --config config              path to config file
+  -s, --symbol symbol              name of the symbol to start the navigation from
+  -j, --output-type type           type of output: graphOnly, jsonOutputPlain, jsonOutputB64 or jsonOutputGZB64 (default "graphOnly")
+  -x, --max-depth number           max number of levels in call flow exploration (0=No limit)
+  -m, --mode mode                  mode of plotting: 1=Symbols, 2=Subsystems, 3=Subsystems with labels, 4=Target subsystem isolation (default 2)
+  -b, --excluded-before symbols    list of symbols to exclude before the target symbol
+  -a, --excluded-after symbols     list of symbols to exclude after the target symbol
+  -t, --target-subsys subsystems   list of subsystems to include in the output
+  -e, --db-driver driver           database driver: mysql, postgres or sqlite3 (default "postgres")
+  -d, --DBDSN DSN                  database DSN (format: "host=dbhost.com port=5432 user=username password=<password> dbname=kernel_bin sslmode=disable")
```
Comment (rokkbert) [inline README.md]: Here we give a format, but later we say "DSN in the engine specific format", which seems more correct. I think we need to give examples for all DB engines we support.

File: config/config.go [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+	"os"
+	"path/filepath"
+
+	"github.com/spf13/pflag"
+	c "nav/constants"
+)
+
+type Config struct {
+	ConfValues ConfValues
+	// add DB connection instance here
```
Comment (rokkbert) [inline config/config.go]: Why, it's part of ConfValues already, isn't it? Or is that something else?

Comment (rokkbert) [reply config/config.go]: Actually, I'm not sure I understand the benefit of this indirection. Why not just use the ConfValues struct but call it Config?

File: config/config.go [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+	confValues, err := initConfig()
+	config.ConfValues = confValues
+	if err != nil {
+		return nil, err
+	}
+	return &config, nil
+}
+
+// initConfig gathers the configuration values from command line and config file (if provided) and returns
+// them in a ConfValues struct. If both are provided, the command line flags take precedence.
```
Comment (rokkbert) [inline config/config.go:44]: Nice comment!

File: README.md [hunk 10]
```diff
@@ -42,46 +42,53 @@ $ ./nav -f conf.json -s kernel_init
...
-|Target_sybsys|List of subsys that need to be highlighted. if empty, only the subs that contain the start is highlighted  |string  |[]                 | 
+| Field           | description                                                                                               | type     | Default value |
+|-----------------|-----------------------------------------------------------------------------------------------------------|----------|---------------|
+| db_driver       | Name of DB engine driver, i.e. postgres, mysql or sqlite3                                                 | string   | postgres      |
+| DBDSN           | DSN in the engine specific format                                                                         | string   | See Note      |
+| db_instance     | Database instance                                                                                         | int      | 1             |
+| symbol          | The interesting symbols instance identifier                                                               | string   | NULL          |
```
Comment (rokkbert) [inline README.md]: This description seems off.

File: config/config.go [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+	fs := pflag.NewFlagSet("nav", pflag.ContinueOnError)
+	var confValues ConfValues
+	var configPath string
+
+	err := parseCommandLine(fs, &configPath, os.Args)
+	if err != nil {
+		return ConfValues{}, fmt.Errorf("error initializing command line flags: %w", err)
+	}
+
+	if len(os.Args) == 1 {
```
Comment (rokkbert) [inline config/config.go]: I'd move this check as a guard clause to the top. Nothing to parse in this case.

File: config/config.go [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+		return ConfValues{}, fmt.Errorf("invalid configuration: %w", err)
+	}
+
+	return confValues, nil
+}
+
+func loadConfigFile(cfg *ConfValues, configPath string) error {
+	cleanedPath := filepath.Clean(configPath)
+
+	if cleanedPath != configPath {
```
Comment (rokkbert) [inline config/config.go:77]: This seems unnecessarily strict to me. Why disallow "unclean" paths? Maybe they are automatically constructed and contain /./ or other redundancies, but why not?

File: constants/const.go [hunk 13]
```diff
@@ -0,0 +1,27 @@
+package constants
```
Comment (rokkbert) [inline constants/const.go:1]: Could the constants from flags.go be here too? What sets them apart?

File: constants/const.go [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+
+const (
+	OSExitSuccess = 0
+	OSExitError   = -1
+)
+
+type OutMode int64
```
Comment (rokkbert) [inline constants/const.go:8]: Why is it int64, if the values go only from 1 to 4?

File: README.md (hunk 7 above)
Comment (alessandrocarminati) [reply README.md]: Agreed, providing examples for all the supported database engines could be helpful for further clarity. 
As for DSN documentation, I have already started documenting it in the kern_bin_db repository's README[1]. 
It might be worth repeating the same information in the nav README for easier reference. 
Please note that the documentation is not exhaustive as I haven't had the chance to test all the DSN, but it still provides useful indications, including for the untested sqlite. 

[1]. https://github.com/AC-FuSa-Tools/kern_bin_db-/blob/main/README.md"

File: config/config.go (hunk 11 above)
Comment (alessandrocarminati) [reply config/config.go]: My initial idea for configuration was to include a default config, the ability to define items using a configuration file, and to accept command line arguments. According to this approach, the working configuration should have been the combination of all three in the order listed, with the default being overwritten by the configuration file and the sum of the two being overwritten by command line arguments, making the latter more powerful than the former two. Admittedly, my own laziness often led me to rely on the default configuration, which was sufficient in most cases. While I understand that having a default for only one user is not necessary, I believe that other users with similar tendencies might appreciate such a feature. Since we cannot provide a custom binary, perhaps we could define a default configuration file name that is used when the binary is called without arguments.

File: config/config.go (hunk 12 above)
Comment (alessandrocarminati) [reply config/config.go:77]: The necessity of path cleaning largely depends on the use case. 
While I have a tendency to overlook such issues, I am convinced that sanitized input is generally beneficial.
Without a defined use case, there may be no difference between clean or unclean paths. 
However, if we consider security concerns, having a cleaned path could be useful for further filtering, and it would be good practice to verify that the path is a regular file. 
Although I agree with Rob that it is unnecessary without a specified use case, I see it as a useful starting point for improving the tool's security. 
It's worth noting that one of our future goals is to promote the tools to become a web service, while still keeping the tools as a command-line interface. 
Therefore, we should address these kinds of issues at some point.

File: constants/const.go (hunk 14 above)
Comment (alessandrocarminati) [reply constants/const.go:8]: This inconsistency with the int64 type comes from the original code and was chosen based on considerations for the architecture that will run the code. When picking a type, the register size was taken into account. On the other hand, the int type was chosen just because it was easy at the time. I agree that it would be better to make the two consistent. Using a 32 or 64 bit int would likely be best for performance, whereas using an 8 bit int may not be ideal. I have not verified how the Go compiler stores bytes in memory, but GCC typically pads 16 and 8 bit integer types to 32 or 64 bits, unless instructed otherwise. While I used to be sensitive to these low-level considerations, I have come to understand that it is better to leave them to the compiler unless there is a serious issue to address. In this case, the only problem I see is the inconsistent types used for basically the same enum.

File: config/config.go [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+		fmt.Printf("No database driver specified. Defaulting to %s.\n", defaultDBDriver)
+		return nil
+	case "mysql", "postgres", "sqlite3":
+		return nil
+	default:
+		return fmt.Errorf("invalid database driver: %s\nChoose one of the following: mysql, postgres or sqlite3", *d)
+	}
+}
+
+func validateMode(m *c.OutMode) error {
```
Comment (alessandrocarminati) [inline config/config.go:150]: So far, the mode argument was a numerical value while the output type was a string. 
Reviewing the repository history, it appears that the mode was implemented prior to the output type. 
When I introduced the output type, I believed that having too many numerical values would make the command line interface confusing, so I opted for a string representation instead to improve readability. 
While it wasn't an urgent matter at the time, I always had the notion that the mode would eventually need to be converted into a string argument as well. 
With our current efforts focused on refactoring the configuration-related aspects, I believe it is an appropriate time to make this conversion.

File: config/config.go (hunk 8 above)
Comment (Darth-Mera) [reply config/config.go]: Currently the DB connection is happening in main():

```
	t := connectToken{conf.ConfValues.DBDriver, conf.ConfValues.DBDSN}
	db := connectDb(&t)

```
My thinking was that it would make more sense if the db connection instance lived together with the configuration logic, and was just passed around the rest of the program as a pointer. So the plan was to move the connection logic to the config package as part of another PR and have the instance here.

File: config/config.go (hunk 12 above)
Comment (Darth-Mera) [reply config/config.go:77]: I think it's always good to validate and sanitize user input. Maybe in the future it will turn out that this kind of sanitization causes more problems than good and will need to be changed, but I think it's better than having none at all.

File: README.md (hunk 7 above)
Comment (Darth-Mera) [reply README.md]: Addressing this in the following commit, therefore I'm marking this as resolved.

File: README.md (hunk 10 above)
Comment (Darth-Mera) [reply README.md]: Addressing this in the following commit, therefore I'm marking this as resolved.

File: config/config.go (hunk 11 above)
Comment (Darth-Mera) [reply config/config.go]: @alessandrocarminati Maybe this is something that could be written down as a Github issue and addressed in some future PR?

File: config/config.go (hunk 15 above)
Comment (Darth-Mera) [reply config/config.go:150]: I think this is also something that can be written down as a Github issue and addressed in a future PR, since I'd like to avoid changing this PR much and complicating its review.

File: constants/const.go (hunk 13 above)
Comment (Darth-Mera) [reply constants/const.go:1]: Addressing this in the following commit, therefore I'm marking this as resolved.

File: config/config.go (hunk 8 above)
Comment (rokkbert) [reply config/config.go]: Ah ok, so that is a TODO for later and "DB connection instance" != "DB instance", as I had thought.

File: config/config.go (hunk 12 above)
Comment (rokkbert) [reply config/config.go:77]: I see. Especially when we start feeding parameters from the web it makes sense to be careful. But then I suggest we give the user more information about the issue, because I would not expect to be an unclean path to be invalid and it would take me time to find the problem. Example (makes only sense if some primitive script generated the path, of course, nobody would type that):
$ ./nav -f ../nav/../nav//example_config.json 
error: invalid file path

nicer:
$ ./nav -f ../nav/../nav//example_config.json 
error: invalid file path, did you mean '../nav/example_config.json'?

File: constants/const.go (hunk 14 above)
Comment (rokkbert) [reply constants/const.go:8]: In the case of individual variables IMHO the size/padding is irrelevant, from a performance PoV, compared to the readability. I think it should be the same type as for the output type consts, because otherwise the reader will wonder what the difference means, even though it doesn't really mean anything.

File: constants/const.go [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+// Configuration defaults.
+const (
+	DefaultMode       = PrintSubsys
+	DefaultOutputType = "graphOnly"
+	DefaultSymbol     = ""
+	DefaultMaxDepth   = 0
+	DefaultDBDriver   = "postgres"
+	DefaultDBInstance = 1
+	DefaultDBDSN      = ""
+)
```
Comment (rokkbert) [inline constants/const.go:36]: What purpose is served by having default values that can't work, like "" for DefaultSymbol and DefaultDBDSN? Reading  initFlagSet it looks like there are defaults for everything, but actually these two have to be defined by the user (command line arguments or config file). Should we remove these constants and just set nil as default value, or something like that? I'm ot experienced enough with pflag to know what the canonical way is.

Comment (Darth-Mera) [reply constants/const.go:36]: This is kind of a leftover from the "reasonable default value for everything" approach that Alessandro argues for. Right now, they indeed don't serve much of a purpose. 
I think it depends on if we want to always have the user to set those (then we can remove the constants) or if we want to have a reasonable default value that the program jumps to if it doesn't receive it (then we can use the constants).

Comment (rokkbert) [reply constants/const.go:36]: I'm very much in favour of reasonable default values, if they exist. I doubt that for these two they do.

File: config/config.go (hunk 8 above)
Comment (Darth-Mera) [reply config/config.go]: I have created an issue for this, closing this conversation as resolved

File: config/config.go (hunk 11 above)
Comment (Darth-Mera) [reply config/config.go]: Opened an issue for this, resolving

File: config/config.go (hunk 12 above)
Comment (Darth-Mera) [reply config/config.go:77]: Addressing in my next commit, marking as resolved

File: constants/const.go (hunk 14 above)
Comment (Darth-Mera) [reply constants/const.go:8]: I've opened an issue for it - I think it's better to get this PR to finally merge, and change it in a new PR. Therefore I'm marking this conversation as resolved.

File: config/config.go (hunk 15 above)
Comment (Darth-Mera) [reply config/config.go:150]: Created an issue for it, closing this conversation

File: constants/const.go (hunk 16 above)
Comment (Darth-Mera) [reply constants/const.go:36]: Removing the redundant constants in my next commit, resolving this conversation

**Comment by rokkbert** [comment]
LGTM now.

File: config/config.go (hunk 8 above)
Comment (Darth-Mera) [reply config/config.go]: https://github.com/AC-FuSa-Tools/nav/issues/25

File: config/config.go (hunk 11 above)
Comment (Darth-Mera) [reply config/config.go]: https://github.com/AC-FuSa-Tools/nav/issues/26

File: constants/const.go (hunk 14 above)
Comment (Darth-Mera) [reply constants/const.go:8]: https://github.com/AC-FuSa-Tools/nav/issues/27

File: config/config.go (hunk 15 above)
Comment (Darth-Mera) [reply config/config.go:150]: https://github.com/AC-FuSa-Tools/nav/issues/27
